File: src/server/node.js
'use strict';
/**
 * A storage node: the server that actually moves the bytes.
 *
 * Concurrency model
 * -----------------
 * Node.js runs ONE thread with an event loop. It does not need a thread per
 * client, because it never blocks waiting for a disk or a socket: it registers
 * "tell me when this chunk is ready" and goes off to serve somebody else.
 * A single node therefore handles many simultaneous GB-sized transfers, and
 * the platform runs three such nodes side by side.
 *
 * See decision.md section 3 for the full argument, including the honest
 * limitation (one node cannot use more than one CPU core for its own work).
 *
 * Run three of them for the demo:
 *     npm run node1        (or: node src/server/node.js 1)
 *     npm run node2
 *     npm run node3
 */

const net = require('net');

const config = require('../common/config');
const {
  bus, WILDCARD, CLIENT_CONNECTED, CLIENT_DISCONNECTED, NODE_STARTED,
} = require('../common/events');
const {
  readJsonLine, sendJson, isDisconnect, PartialTransferError,
} = require('../common/protocol');
const { OPS } = require('./handlers');
const { Monitor } = require('./monitor');
const { PolicyStore } = require('./policy');
const { Storage } = require('./storage');

// Events that are too chatty to print on the console.
const QUIET_EVENTS = ['upload.progress', 'download.progress'];

/** Everything a handler needs, plus the live connection counters. */
class NodeContext {
  constructor(nodeId, storage, policy) {
    this.nodeId = nodeId;
    this.storage = storage;
    this.policy = policy;
    this._active = 0;
    this._served = 0;
  }

  beginRequest() {
    this._active += 1;
    this._served += 1;
  }

  endRequest() {
    this._active -= 1;
  }

  activeCount() {
    return this._active;
  }

  servedCount() {
    return this._served;
  }
}

/** A second bus listener, kept separate from the monitor on purpose. */
function consoleLogger(event) {
  if (QUIET_EVENTS.includes(event.type)) return;

  const stamp = new Date(event.ts * 1000).toLocaleTimeString();
  const details = Object.entries(event)
    .filter(([key, value]) => key !== 'type' && key !== 'ts' && value !== undefined)
    .map(([key, value]) => `${key}=${value}`)
    .join(' ');
  console.log(`[${stamp}] ${event.type.padEnd(20)} ${details}`);
}

/** Handle one client connection from its first byte to its last. */
async function handleConnection(ctx, socket) {
  const peer = `${socket.remoteAddress}:${socket.remotePort}`;
  let counted = false;

  // A socket with no 'error' listener throws and kills the process. A client
  // that is killed mid-upload sends a reset, so this is not a rare case.
  socket.on('error', () => {});

  try {
    const request = await readJsonLine(socket);
    if (!request) return; // peer closed without asking for anything

    const operation = String(request.op || '').toUpperCase();

    // Health pings are bookkeeping, not user load: they must not show up in
    // the number the balancer uses to compare nodes.
    if (operation !== 'PING') {
      ctx.beginRequest();
      counted = true;
      bus.publish(CLIENT_CONNECTED, { peer, node: ctx.nodeId });
    }

    const handler = OPS[operation];
    if (!handler) {
      sendJson(socket, { ok: false, error: `unknown operation: ${operation}` });
    } else {
      await handler(ctx, request, socket);
    }
  } catch (err) {
    // Abrupt client death: the handlers already published the abort event.
    if (!isDisconnect(err) && !(err instanceof PartialTransferError)) {
      console.error(`[${ctx.nodeId}] request failed:`, err.message);
    }
  } finally {
    if (counted) {
      ctx.endRequest();
      bus.publish(CLIENT_DISCONNECTED, { peer, node: ctx.nodeId });
    }
    socket.end();
  }
}

function main() {
  const index = Number(process.argv[2]) - 1;
  if (!Number.isInteger(index) || index < 0 || index >= config.NODES.length) {
    console.log(`usage: node src/server/node.js <node-number 1..${config.NODES.length}>`);
    process.exit(1);
  }

  config.ensureDirs();
  const { host, port } = config.NODES[index];
  const nodeId = `node${index + 1}`;

  const storage = new Storage();
  const policy = new PolicyStore();
  const ctx = new NodeContext(nodeId, storage, policy);

  const monitor = new Monitor(nodeId, `${host}:${port}`);
  monitor.attachActiveSource(() => ctx.activeCount());
  bus.subscribe(WILDCARD, consoleLogger);

  const server = net.createServer((socket) => handleConnection(ctx, socket));

  server.listen(port, host, () => {
    bus.publish(NODE_STARTED, { node: nodeId, address: `${host}:${port}` });
    console.log(`${nodeId} listening on ${host}:${port}  (storage: ${storage.root})`);
  });

  server.on('error', (err) => {
    console.error(`${nodeId} could not start:`, err.message);
    process.exit(1);
  });

  const shutdown = () => {
    console.log(`\n${nodeId} shutting down`);
    monitor.removeStatusFile();
    server.close(() => process.exit(0));
    setTimeout(() => process.exit(0), 1000).unref(); // do not hang on open sockets
  };
  process.on('SIGINT', shutdown);
  process.on('SIGTERM', shutdown);
}

if (require.main === module) main();

module.exports = { NodeContext, handleConnection };

File: src/server/node.py
"""A storage node: the multithreaded server that actually moves the bytes.

Threading model
---------------
One acceptor thread waits for connections. Every accepted connection is given
its own worker thread (socketserver.ThreadingTCPServer), so a 5 GB upload from
one user never blocks a small LIST from another.

Run three of them for the demo:
    python -m src.server.node 1
    python -m src.server.node 2
    python -m src.server.node 3
"""

import socketserver
import sys
import threading
import time

from ..common import config
from ..common.events import (bus, WILDCARD, CLIENT_CONNECTED, CLIENT_DISCONNECTED,
                             NODE_STARTED)
from ..common.protocol import recv_json, send_json
from .handlers import OPS
from .monitor import Monitor
from .policy import PolicyStore
from .storage import Storage

# Events that are too chatty to print on the console.
QUIET_EVENTS = ("upload.progress", "download.progress")


class NodeContext:
    """Everything a handler needs, plus the live connection counters."""

    def __init__(self, node_id, storage, policy):
        self.node_id = node_id
        self.storage = storage
        self.policy = policy
        self._active = 0
        self._served = 0
        self._lock = threading.Lock()

    def begin_request(self):
        with self._lock:
            self._active += 1
            self._served += 1

    def end_request(self):
        with self._lock:
            self._active -= 1

    def active_count(self):
        with self._lock:
            return self._active

    def served_count(self):
        with self._lock:
            return self._served


class ClientHandler(socketserver.BaseRequestHandler):
    """Runs in its own thread. Reads one request and dispatches it."""

    def handle(self):
        ctx = self.server.ctx
        peer = "%s:%d" % self.client_address
        counted = False

        stream = self.request.makefile("rb")
        try:
            request = recv_json(stream)
            if request is None:
                return
            operation = str(request.get("op", "")).upper()

            # Health pings are bookkeeping, not user load: they must not show
            # up in the number the balancer uses to compare nodes.
            if operation != "PING":
                ctx.begin_request()
                counted = True
                bus.publish(CLIENT_CONNECTED, peer=peer, node=ctx.node_id)

            handler = OPS.get(operation)
            if handler is None:
                send_json(self.request, {"ok": False,
                                         "error": "unknown operation: %s" % operation})
            else:
                handler(ctx, request, self.request, stream)
        except (ConnectionError, OSError, ValueError):
            # Abrupt client death. Handlers already published the abort event.
            pass
        finally:
            try:
                stream.close()
            except OSError:
                pass
            if counted:
                ctx.end_request()
                bus.publish(CLIENT_DISCONNECTED, peer=peer, node=ctx.node_id)


class StorageNode(socketserver.ThreadingTCPServer):
    daemon_threads = True      # worker threads never block shutdown
    allow_reuse_address = True

    def __init__(self, address, ctx):
        socketserver.ThreadingTCPServer.__init__(self, address, ClientHandler)
        self.ctx = ctx


def console_logger(event):
    """A second bus listener, kept separate from the monitor on purpose."""
    if event["type"] in QUIET_EVENTS:
        return
    stamp = time.strftime("%H:%M:%S", time.localtime(event["ts"]))
    details = " ".join("%s=%s" % (k, v) for k, v in event.items()
                       if k not in ("type", "ts") and v is not None)
    print("[%s] %-20s %s" % (stamp, event["type"], details))


def main():
    if len(sys.argv) < 2:
        print("usage: python -m src.server.node <node-number 1..%d>" % len(config.NODES))
        return
    index = int(sys.argv[1]) - 1
    if not 0 <= index < len(config.NODES):
        print("no such node number")
        return

    config.ensure_dirs()
    address = config.NODES[index]
    node_id = "node%d" % (index + 1)

    storage = Storage()
    policy = PolicyStore()
    ctx = NodeContext(node_id, storage, policy)

    monitor = Monitor(node_id, storage, address)
    monitor.attach_active_source(ctx.active_count)
    bus.subscribe(WILDCARD, console_logger)

    server = StorageNode(address, ctx)
    bus.publish(NODE_STARTED, node=node_id, address="%s:%d" % address)
    print("%s listening on %s:%d  (storage: %s)" % (node_id, address[0], address[1],
                                                    storage.root))
    try:
        server.serve_forever()
    except KeyboardInterrupt:
        print("\n%s shutting down" % node_id)
    finally:
        server.shutdown()
        server.server_close()
        monitor.remove_status_file()


if __name__ == "__main__":
    main()
